presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedNoAggregationScalarToJoin.java

```diff
@@ -0,0 +1,176 @@
+
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.metadata.FunctionRegistry;
+import com.facebook.presto.spi.type.BooleanType;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.optimizations.PlanNodeDecorrelator;
+import com.facebook.presto.sql.planner.optimizations.PlanNodeDecorrelator.DecorrelatedNode;
+import com.facebook.presto.sql.planner.optimizations.ScalarSubqueryToJoinRewriter;
+import com.facebook.presto.sql.planner.optimizations.ScalarSubqueryToJoinRewriter.SubqueryEquivalentJoin;
+import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.LateralJoinNode;
+import com.facebook.presto.sql.planner.plan.MarkDistinctNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.Cast;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.SimpleCaseExpression;
+import com.facebook.presto.sql.tree.StringLiteral;
+import com.facebook.presto.sql.tree.WhenClause;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+import static com.facebook.presto.matching.Pattern.nonEmpty;
+import static com.facebook.presto.spi.StandardErrorCode.SUBQUERY_MULTIPLE_ROWS;
+import static com.facebook.presto.spi.type.StandardTypes.BOOLEAN;
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher.searchFrom;
+import static com.facebook.presto.sql.planner.optimizations.QueryCardinalityUtil.isScalar;
+import static com.facebook.presto.sql.planner.plan.Patterns.LateralJoin.correlation;
+import static com.facebook.presto.sql.planner.plan.Patterns.lateralJoin;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.util.MorePredicates.isInstanceOfAny;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Scalar filter scan query is something like:
+ * <pre>
+ *     SELECT a,b,c FROM rel WHERE a = correlated1 AND b = correlated2
+ * </pre>
+ * <p>
+ * This optimizer can rewrite to aggregation over a left outer join:
+ * <p>
+ * From:
+ * <pre>
+ * - LateralJoin (with correlation list: [C])
+ *   - (input) plan which produces symbols: [A, B, C]
+ *   - (scalar subquery) Project F
+ *     - Filter(D = C AND E > 5)
+ *       - plan which produces symbols: [D, E, F]
+ * </pre>
+ * to:
+ * <pre>
+ * - Filter(CASE isDistinct WHEN true THEN true ELSE fail('Scalar sub-query has returned multiple rows'))
+ *   - MarkDistinct(isDistinct)
+ *     - Join(LEFT_OUTER, D = C)
```

Comment:
Do we really need a direct rewrite to Join? Without it the code would be much simpler. What do you think?

Please see my comment: https://github.com/prestodb/presto/pull/8435#discussion_r181747306

Reply:
: )

Please see my comment: https://github.com/prestodb/presto/pull/8435#discussion_r181637848

I cannot rewrite to lateral join as lateral join do not support join condition and I have correlated predicates which I have to put it into some kind of a join. In case of inner join I could add filter on top, but this a left join so I cannot.

Reply:
Let's discuss offline. 

> do not support join condition and I have correlated predicates which I have to put it into some kind of a join

My point is that you don't have to decorrelated subquery at all.

Reply:
Finally, got it. Thank you!